env0/templates.go

```diff
@@ -226,3 +228,89 @@ func getTemplateSchema(templateType TemplateType) map[string]*schema.Schema {
 
 	return s
 }
+
+func templateCreatePayloadRetryOnHelper(prefix string, d *schema.ResourceData, retryType string, retryOnPtr **client.TemplateRetryOn) {
+	if prefix != "" {
+		prefix += "."
+	}
+	retries, hasRetries := d.GetOk(prefix + "retries_on_" + retryType)
+	if hasRetries {
+		retryOn := &client.TemplateRetryOn{
+			Times: retries.(int),
+		}
+		if retryIfMatchesRegex, ok := d.GetOk(prefix + "retry_on_" + retryType + "_only_when_matches_regex"); ok {
+			retryOn.ErrorRegex = retryIfMatchesRegex.(string)
+		}
+
+		*retryOnPtr = retryOn
+	}
+}
+
+func templateCreatePayloadFromParameters(prefix string, d *schema.ResourceData) (client.TemplateCreatePayload, diag.Diagnostics) {
+	var payload client.TemplateCreatePayload
+	if err := readResourceDataEx(prefix, &payload, d); err != nil {
+		return payload, diag.Errorf("schema resource data serialization failed: %v", err)
+	}
+
+	tokenIdKey := "token_id"
+	if prefix != "" {
+		tokenIdKey = prefix + "." + tokenIdKey
+	}
+	if tokenId, ok := d.GetOk(tokenIdKey); ok {
+		payload.IsGitLab = tokenId != ""
+	}
+
+	templateCreatePayloadRetryOnHelper(prefix, d, "deploy", &payload.Retry.OnDeploy)
+	templateCreatePayloadRetryOnHelper(prefix, d, "destroy", &payload.Retry.OnDestroy)
+
+	if err := payload.Validate(); err != nil {
+		return payload, diag.Errorf(err.Error())
+	}
+
+	return payload, nil
+}
+
+// Reads template and writes to the resource data.
+func templateRead(prefix string, template client.Template, d *schema.ResourceData) error {
+	if prefix != "" {
+		templates := []client.Template{template}
+		if err := writeResourceDataSlice(templates, prefix, d); err != nil {
+			return fmt.Errorf("schema resource data serialization failed: %v", err)
+		}
+	} else if err := writeResourceData(&template, d); err != nil {
+		return fmt.Errorf("schema resource data serialization failed: %v", err)
+	}
+
+	templateReadRetryOnHelper(prefix, d, "deploy", template.Retry.OnDeploy)
+	templateReadRetryOnHelper(prefix, d, "destroy", template.Retry.OnDestroy)
+
+	return nil
+}
+
+// Helpers function for templateRead.
+func templateReadRetryOnHelper(prefix string, d *schema.ResourceData, retryType string, retryOn *client.TemplateRetryOn) {
+	if prefix != "" {
+		values := d.Get(prefix)
```

Comment:
There isn't much I can do about this ugliness.
Set("template.0...") isn't allowed.